internal/topology/check/compatibility.go

```diff
@@ -54,3 +62,261 @@ func ObjectsAreInTheSameNamespace(current, desired client.Object) error {
+	if err := TemplatesAreCompatible(in.Spec.ControlPlane.LocalObjectTemplate, old.Spec.ControlPlane.LocalObjectTemplate); err != nil {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "controlPlane", "localObjectTemplate", "ref"),
+				in.Spec.ControlPlane.LocalObjectTemplate.Ref,
+				err.Error()),
+		)
+	}
+
+	if in.Spec.ControlPlane.MachineInfrastructure != nil && old.Spec.ControlPlane.MachineInfrastructure != nil {
+		if err := TemplatesAreCompatible(*in.Spec.ControlPlane.MachineInfrastructure, *old.Spec.ControlPlane.MachineInfrastructure); err != nil {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("spec", "controlPlane", "machineInfrastructure", "ref"),
+					in.Spec.ControlPlane.MachineInfrastructure.Ref,
+					err.Error()),
+			)
+		}
+	}
+
+	// Validate changes to MachineDeployments.
+	allErrs = append(allErrs, MachineDeploymentClassesAreCompatible(old, in)...)
+
+	return allErrs
+}
+
+// MachineDeploymentClassesAreCompatible checks if each MachineDeploymentClass in the new ClusterClass is a compatible change from the previous ClusterClass.
+// It checks if:
+// 1) Any MachineDeploymentClass has been removed.
+// 2) If the MachineDeploymentClass.Template.Infrastructure reference has changed its Group or Kind.
+func MachineDeploymentClassesAreCompatible(old, in *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// Ensure no MachineDeployment class was removed.
+	classes := classNamesFromWorkerClass(in.Spec.Workers)
+	for _, oldClass := range old.Spec.Workers.MachineDeployments {
+		if !classes.Has(oldClass.Class) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "workers", "machineDeployments"),
+					in.Spec.Workers.MachineDeployments,
+					fmt.Sprintf("The %q MachineDeployment class can't be removed.", oldClass.Class),
+				),
+			)
+		}
+	}
+
+	// Ensure previous MachineDeployment class was modified in a compatible way.
+	for i, class := range in.Spec.Workers.MachineDeployments {
+		for _, oldClass := range old.Spec.Workers.MachineDeployments {
+			if class.Class == oldClass.Class {
+				// NOTE: class.Template.Metadata and class.Template.Bootstrap are allowed to change;
+				// class.Template.Bootstrap are ensured syntactically correct by validateAllRefs.
+				// Validates class.Template.Infrastructure template changes in a compatible way
+				if err := TemplatesAreCompatible(class.Template.Infrastructure, oldClass.Template.Infrastructure); err != nil {
+					allErrs = append(allErrs,
+						field.Invalid(field.NewPath("spec", "workers", "machineDeployments").Index(i).Child("template", "infrastructure"),
+							class.Template.Infrastructure.Ref,
+							err.Error(),
+						),
+					)
+				}
+			}
+		}
+	}
+	return allErrs
+}
+
+// MachineDeploymentClassesAreUnique checks that no two MachineDeploymentClasses in a ClusterClass share a name.
+func MachineDeploymentClassesAreUnique(in *clusterv1.ClusterClass) field.ErrorList {
```

Comment:
I would personally just consistently use:
`cluster`, `oldCluster`, `newCluster`, `clusterClass` etc..

I think the current mix is not super great :)
(I'm aware we had this problem before)